pkg/asset/kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,157 @@
+package kubeconfig
+
+import (
+	"encoding/base64"
+	"fmt"
+	"path/filepath"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/types"
+	clientcmd "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+const (
+	// KubeconfigUserNameAdmin is the user name of the admin kubeconfig.
+	KubeconfigUserNameAdmin = "admin"
+	// KubeconfigUserNamekubelet is the user name of the kubelet kubeconfig.
+	KubeconfigUserNamekubelet = "kubelet"
+)
+
+// Kubeconfig implements the asset.Asset interface that generates
+// the admin kubeconfig and kubelet kubeconfig.
+type Kubeconfig struct {
+	rootDir       string
+	userName      string // admin or kubelet.
+	rootCA        asset.Asset
+	certKey       asset.Asset
+	installConfig asset.Asset
+}
+
+var _ asset.Asset = (*Kubeconfig)(nil)
+
+// Dependencies returns the dependency of the kubeconfig.
+func (k *Kubeconfig) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		k.rootCA,
+		k.certKey,
+		k.installConfig,
+	}
+}
+
+// Generate generates the kubeconfig.
+func (k *Kubeconfig) Generate(parents map[asset.Asset]*asset.State) (*asset.State, error) {
+	var err error
+
+	caCertData, err := getCertKeyData(k.rootCA, parents, ".crt")
+	if err != nil {
+		return nil, err
+	}
+	clientKeyData, err := getCertKeyData(k.certKey, parents, ".key")
+	if err != nil {
+		return nil, err
+	}
+	clientCertData, err := getCertKeyData(k.certKey, parents, ".crt")
+	if err != nil {
+		return nil, err
+	}
+	installConfig, err := getInstallConfig(k.installConfig, parents)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfig := clientcmd.Config{
+		Clusters: []clientcmd.NamedCluster{
+			{
+				Name: installConfig.Name,
+				Cluster: clientcmd.Cluster{
+					Server: fmt.Sprintf("https://%s-api.%s:6443", installConfig.Name, installConfig.BaseDomain),
+					CertificateAuthorityData: caCertData,
+				},
+			},
+		},
+		AuthInfos: []clientcmd.NamedAuthInfo{
+			{
+				Name: k.userName,
+				AuthInfo: clientcmd.AuthInfo{
+					ClientCertificateData: clientCertData,
+					ClientKeyData:         clientKeyData,
+				},
+			},
+		},
+		Contexts: []clientcmd.NamedContext{
+			{
+				Name: k.userName,
+				Context: clientcmd.Context{
+					Cluster:  installConfig.Name,
+					AuthInfo: k.userName,
+				},
+			},
+		},
+		CurrentContext: k.userName,
+	}
+
+	data, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	st := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: assetFilePath(k.rootDir, k.userName),
+				Data: data,
+			},
+		},
+	}
+
+	if err := st.PersistToFile(); err != nil {
+		return nil, err
+	}
+
+	return st, nil
+}
+
+func assetFilePath(rootDir, userName string) string {
+	if userName == KubeconfigUserNamekubelet {
+		return filepath.Join(rootDir, "auth", "kubeconfig-kubelet")
+	}
+	return filepath.Join(rootDir, "auth", "kubeconfig")
```

Comment:
Why is the admin kubeconfig not getting namespaced?  I'd have expected we could use:

```go
filepath.Join(rootDir, "auth", fmt.Sprintf("kubeconfig-%s", k.userName))
```

for both the admin and kublet cases, in which case we can drop the `assetFilePath` helper.

Reply:
This is equivalent to what we have now, so I want to keep the same convention for old folks.

Reply:
Actually I updated the filename.